### PR TITLE
Add NTS support

### PR DESCRIPTION
**Pull request by mlichvar**
These patches add support for the client Network Time Security (NTS) option in chrony.

### CODE REVIEW

**Comment by mlichvar** [comment]
This doesn't currently work well with the "one_time_sync" that's performed after enabling the NTP switch. As ntplib doesn't support NTS, the function would need to call `chronyd -q 'server {server} nts iburst maxsamples 1' to make a quick step using an NTS-enabled server.

Maybe the one_time_sync code could be dropped and just rely on the restarted chronyd service to make the step? It should take about 6-8 seconds. Is that too long?

**Comment by poncovka** [comment]
Jenkins, it's ok to test.

File: pyanaconda/ntp.py [hunk 1]
```diff
@@ -106,7 +110,12 @@ def ntp_server_working(server_hostname):
     client = ntplib.NTPClient()
```
Comment (M4rtinK) [inline pyanaconda/ntp.py:110]: I think we might move this in the else branch as the `client` variable is now used only there ?

File: pyanaconda/ntp.py [hunk 2]
```diff
@@ -143,7 +152,16 @@ def get_servers_from_config(conf_file_path=NTP_CONFIG_FILE):
...
-                server.options = ["iburst"]
+                server.options = []
+
+                words = match.group(3).lower().split()
+                for i in range(len(words)):
+                    if words[i] in SRV_NOARG_OPTIONS:
+                        server.options.append(words[i])
+                    elif words[i] in SRV_ARG_OPTIONS and i + 1 < len(words):
+                        server.options.append(' '.join(words[i:i+2]))
+                    # ignore unknown options
```
Comment (M4rtinK) [inline pyanaconda/ntp.py:163]: Maybe we can log those unknown options ? That would also make the comment look better and not a bit dangling like this.

File: pyanaconda/ntp.py [hunk 3]
```diff
@@ -106,7 +110,12 @@ def ntp_server_working(server_hostname):
     client = ntplib.NTPClient()
 
     try:
-        client.request(server_hostname)
+        # ntplib doesn't support NTS
+        if "nts" in options:
+            s = socket.create_connection((server_hostname, 4460), 2)
```
Comment (poncovka) [inline pyanaconda/ntp.py:115]: How can I test it? I wasn't able to find a server that would be marked as working.

**Comment by poncovka** [comment]
Here are screenshots of the UI changes:

Screenshot_vpn_test_03_2020-07-16_16:42:48-nts1

Screenshot_vpn_test_03_2020-07-16_16:43:08-nts2

File: pyanaconda/ntp.py [hunk 4]
```diff
@@ -316,7 +334,7 @@ def check_status(self, server):
         threadMgr.add(AnacondaThread(
             prefix=THREAD_NTP_SERVER_CHECK,
             target=self._check_status,
-            args=(hostname, ))
+            args=(hostname, server.options))
```
[the remark this answers is not in the document]
Comment (mlichvar) [reply pyanaconda/ntp.py]: Ok, thanks.

File: pyanaconda/ntp.py (hunk 2 above)
Comment (mlichvar) [reply pyanaconda/ntp.py]: Makes sense. Thanks.

File: pyanaconda/ntp.py (hunk 3 above)
Comment (mlichvar) [reply pyanaconda/ntp.py:113]: You can test it with `nts-test.strangled.net` and `ntp1.glypnod.com` for example. Cloudflare and others will hopefully move to the new port soon. Please note that we will also need to update the chrony package to use this port by default.

**Comment by poncovka** [comment]
> This doesn't currently work very well with the "one_time_sync" that's performed after enabling the NTP switch. An NTS-enabled server may have the NTP server running on different host or port (that's negotiated in NTS-KE), so the ntplib request may not reach the NTP server, and even if it does, the response won't be protected by NTS. I'm not sure how much that matters, as the installer seems to sync to a pool.ntp.org server before the servers are configured in the dialog.
> 
> For a proper NTS-enabled sync, the one_time_sync function would need to call something like `chronyd -q 'server {server} nts iburst maxsamples 1'`.
> 
> But maybe it could be dropped and just rely on the restarted chronyd service to make the step? It should take about 6-8 seconds. Is that too long?

It looks like the logic was introduced in the commit https://github.com/rhinstaller/anaconda/commit/edad0eb1504bde987d2bda7d8ca225266629b81d. It is from 2012, so I don't really know how much time it saved back then. I guess that 6-8 seconds could be fine. @jkonecny12, what do you think?

**Comment by jkonecny12** [comment]
> This doesn't currently work very well with the "one_time_sync" that's performed after enabling the NTP switch. An NTS-enabled server may have the NTP server running on different host or port (that's negotiated in NTS-KE), so the ntplib request may not reach the NTP server, and even if it does, the response won't be protected by NTS. I'm not sure how much that matters, as the installer seems to sync to a pool.ntp.org server before the servers are configured in the dialog.
> 
> For a proper NTS-enabled sync, the one_time_sync function would need to call something like `chronyd -q 'server {server} nts iburst maxsamples 1'`.
> 
> But maybe it could be dropped and just rely on the restarted chronyd service to make the step? It should take about 6-8 seconds. Is that too long?

I just wonder, I remember that chrony when started on system with bad clocks it starts slowly convergent to the correct time but not immediately. Is that still true? In that case will the service restart work at all?

**Comment by mlichvar** [comment]
It depends on the configuration. The default configuration always allowed step on the service (re)start, but it takes at least 6 seconds for that to happen, which I suspect may be confusing when the clock is obviously wrong and the user is staring at the large time displayed in the timezone dialog?

**Comment by mlichvar** [comment]
Thanks for the review.

**Comment by poncovka** [comment]
Tested. The changes for the `one_time_sync` function can be done in another pull request, so I am going to merge this one. I like the idea of dropping the function and relying on the chronyd service.

**Pull request merged**